neo/Network/LocalNode.cs

```diff
@@ -277,38 +278,77 @@ private static void CheckMemPool()
                         }
-                        tasks = endpoints.Select(p => ConnectToPeerAsync(p)).ToArray();
+                        tasksList.AddRange(endpoints.Select(p => ConnectToPeerAsync(p)));
                     }
-                    else if (connectedCount > 0)
+
+                    if (connectedCount > 0)
                     {
-                        lock (connectedPeers)
+                        if (unconnectedCount + connectedCount < DesiredAvailablePeers)
+                        {
+                            lock (connectedPeers)
+                            {
+                                foreach (RemoteNode node in connectedPeers)
+                                    node.RequestPeers();
+                            }
+
+                            if (lastSufficientPeersTimestamp < DateTime.UtcNow.AddSeconds(-180))
+                            {
+                                Random rand = new Random();
+                                tasksList.AddRange(Settings.Default.SeedList
+                                    .OrderBy(p => rand.Next()).Take(2)
+                                    .OfType<string>().Select(p => p.Split(':'))
+                                    .Select(p => ConnectToPeerAsync(p[0], int.Parse(p[1]))));
+                                
+                                // Reset this so we try to get more peers from the seed nodes before trying to connect
+                                // to seed nodes again.
+                                lastSufficientPeersTimestamp = DateTime.UtcNow;
+                            }
+                        }
+                        else
                         {
-                            foreach (RemoteNode node in connectedPeers)
-                                node.RequestPeers();
+                            lastSufficientPeersTimestamp = DateTime.UtcNow;
                         }
                     }
                     else
                     {
                         Random rand = new Random();
-                        tasks = Settings.Default.SeedList.OrderBy(p => rand.Next()).Take(5).OfType<string>().Select(p => p.Split(':')).Select(p => ConnectToPeerAsync(p[0], int.Parse(p[1]))).ToArray();
+                        tasksList.AddRange(Settings.Default.SeedList
+                            .OrderBy(p => rand.Next()).Take(5)
+                            .OfType<string>().Select(p => p.Split(':'))
+                            .Select(p => ConnectToPeerAsync(p[0], int.Parse(p[1]))));
+			// Reset this since we have connected to seed nodes.
+                        lastSufficientPeersTimestamp = DateTime.UtcNow;
                     }
+                    
                     try
                     {
-                        Task.WaitAll(tasks, cancellationTokenSource.Token);
+                        var tasksArray = tasksList.ToArray();
+                        Task.WaitAny(tasksArray, 5000, cancellationTokenSource.Token);
```

Comment:
The method `Task.WaitAny` waits for any of the provided `Task` objects to complete. And it may start a new round of tasks before all the tasks are complete.

Reply:
That is exactly what we want the code to do. It leaves the existing tasks running and when one completes it replaces it with a new task by connecting to another peer if there are less than the desired amount of connections open.

Reply:
Note that it still cleans up all the tasks whenever they complete, because it continues tracking the tasks that are still running when WaitAny returns.

Reply:
Is it possible to add duplicate tasks in the second round of loops, such as connecting to the nodes in `SeedList`?

Reply:
When you say "duplicate tasks" do you mean create a new tasks that is connecting to the same node already connected to?

That is possible, however, the code has a check during connecting that will cause a second task that attempts to connect to the same node to exit during the connection process and then that task will be cleaned up, while the other task that was already connected to that node will just happily continue, so I don't think it is an issue.

Note that these changes will work best along with the code in my other pull request:
https://github.com/neo-project/neo/pull/218

This is because staying connected to more peers means getting more duplicate relayed transactions, and my code in that pull request contains improvements to not repeat verification of the exact same transactions at the same block height.

Reply:
Ok, further testing shows that it does need extra logic to prevent duplicating tasks. I've got a fix I'll be pushing to the branch shortly.

Reply:
I’m a bit delayed on being able to update this pull request, but I’m still planning to do it soon (within a day or two).

Reply:
I pushed updated code so it isn't possible to duplicate tasks in subsequent rounds of loops.